### PR TITLE
[codex] Retry transient remote plugin catalog GETs

### PR DESCRIPTION
**Pull request by aphonpra-oai**
## Summary

- Add a one-shot retry wrapper for idempotent remote plugin catalog GET requests.
- Retry each catalog read independently, including `/ps/plugins/list`, `/ps/plugins/installed`, `/ps/plugins/workspace/shared`, and the created workspace plugin list at `/ps/plugins/workspace/created`.
- Also covers related read-only plugin detail, skill detail, and suggested-plugin GETs that use the same catalog sender.
- Retry only transient transport/body failures plus retryable HTTP statuses: `408`, `429`, and `5xx`.
- Keep state-changing POST mutations such as `/ps/plugins/{id}/install` and `/uninstall` on the existing single-attempt path.

## Why

We saw some missing-plugin reports recover after a top-right refresh or full Codex restart. Both actions trigger fresh plugin discovery, which suggests at least one failure class is transient: a later catalog read can succeed even though the first read failed.

This PR moves a small part of that recovery earlier in the same plugin load. Instead of requiring the user to refresh or restart when one remote catalog GET hits a short-lived failure, Codex retries that individual GET once before surfacing the error.

This is intentionally a conservative mitigation. It can help when plugin-service or a dependency has a brief blip, but it does not fix sustained backend throttling, capacity issues, stale auth/link state, or cases where the authoritative tool listing omits a connector.

## Retry behavior

- Maximum attempts: `2` total per catalog GET: the original request plus one retry.
- Retry scope: each GET retries independently. A transient `/ps/plugins/list` failure does not force `/ps/plugins/installed` or `/ps/plugins/workspace/shared` to retry, and vice versa.
- Retry delay: the default retry waits `250ms`.
- Retryable request errors: reqwest timeout, connect, and body-read errors.
- Retryable HTTP statuses: `408`, `429`, and `5xx`.
- `Retry-After`: numeric `Retry-After` is honored only when it is within the one-shot retry budget. Values longer than `250ms`, or unparsable values, skip the retry so Codex does not add load while the server is asking clients to back off.
- Body-read failures after headers: if a retryable status and `Retry-After` header arrive but the body read later fails, the parsed `Retry-After` is preserved for the retry decision.
- Mutations: install, uninstall, publish, update, and other POST paths still use the existing single-attempt sender and are not retried.

## Out of scope

- This does not implement partial/degraded catalog rendering if one source fails after retry.
- This does not change backend connector authorization, link selection, or `/public/mcp` tool-list behavior.
- This does not refresh local app/tool cache state after an external reauth flow.

## Commit Notes

- Added `send_and_decode_idempotent_get_with_retry` for read-only remote plugin catalog GETs.
- Preserved the existing `send_and_decode` path for POST mutations.
- Propagated body-read failures as request errors so stalled, truncated, or dropped response bodies can use the retry classifier.
- Routed `/ps/plugins/workspace/created` through the retry helper.
- Added bounded `Retry-After` handling for `429` and `5xx`.
- Preserved retryable status/header information when the response body read fails.

## Review Notes

- Addressed @codex feedback:
  - retry body-read timeout/connect/body errors instead of converting them into decode errors
  - avoid retrying inside a longer/unparsable `Retry-After` throttle window
  - route the created workspace plugin list GET through the retry helper
  - preserve `Retry-After` when a retryable status response fails while reading the body

## Validation

- `cargo test -p codex-core-plugins remote_plugin` — 27 passed
- `cargo test -p codex-core-plugins fetch_remote_marketplaces_retries` — 2 passed

### CODE REVIEW

**Comment by aphonpra-oai** [comment]
@codex please review this PR for correctness and edge cases.

Focus areas:
- retry is limited to idempotent remote plugin catalog GETs
- `/list`, `/installed`, and `/workspace/shared` retry independently
- POST install/uninstall mutations are not retried
- retry predicate is appropriately narrow: timeout/connect, `408`, `429`, and `5xx`
- tests cover transient workspace `/installed` and `/list` recovery

**Comment by aphonpra-oai** [comment]
@codex please re-review the latest commit. I addressed the first review feedback by preserving body-read request errors for retry, bounding 429 Retry-After behavior, and routing the created workspace plugin list GET through the retry helper.

**Comment by aphonpra-oai** [comment]
@codex please review current head 0066f87. The previous comments are addressed in this commit: body-read errors now propagate as request errors for retry classification, 429 Retry-After is bounded/skipped, and /ps/plugins/workspace/created uses the retry helper.

**Comment by aphonpra-oai** [comment]
@codex review

**Comment by aphonpra-oai** [comment]
@codex review

**Comment by aphonpra-oai** [comment]
@codex review

Latest head addresses the remaining Retry-After edge case: if a retryable status delivers headers with Retry-After but the response body read fails/truncates, the parsed Retry-After is preserved and the bounded retry gate is still honored. Added a regression test with a truncated 503 + Retry-After: 60 response to prove no second GET is sent.

File: codex-rs/core-plugins/src/remote.rs [hunk 1]
```diff
@@ -1840,25 +1844,135 @@ async fn send_and_decode<T: for<'de> Deserialize<'de>>(
...
+}
+
+fn retry_delay_for_remote_plugin_catalog_get_error(
+    err: &RemotePluginCatalogError,
+    retry_after: RetryAfterDelay,
+) -> Option<Duration> {
+    match err {
+        RemotePluginCatalogError::Request { source, .. } => (source.is_timeout()
+            || source.is_connect())
+        .then_some(REMOTE_PLUGIN_CATALOG_GET_RETRY_DELAY),
```
[the remark this answers is not in the document]
Comment (aphonpra-oai) [reply codex-rs/core-plugins/src/remote.rs]: Addressed in commit 6be86c2. The retry predicate now includes reqwest body read errors (`source.is_body()`), so transient dropped/truncated response bodies are covered by the one-shot retry.

File: codex-rs/core-plugins/src/remote.rs [hunk 2]
```diff
@@ -1840,25 +1844,135 @@ async fn send_and_decode<T: for<'de> Deserialize<'de>>(
...
+                RetryAfterDelay::Delay(delay) if delay <= REMOTE_PLUGIN_CATALOG_GET_RETRY_DELAY => {
+                    Some(delay)
+                }
+                RetryAfterDelay::Delay(_) | RetryAfterDelay::Unparsable => None,
+                RetryAfterDelay::None => Some(REMOTE_PLUGIN_CATALOG_GET_RETRY_DELAY),
+            }
+        }
+        RemotePluginCatalogError::UnexpectedStatus { status, .. } => {
+            (*status == reqwest::StatusCode::REQUEST_TIMEOUT || status.is_server_error())
+                .then_some(REMOTE_PLUGIN_CATALOG_GET_RETRY_DELAY)
```
[the remark this answers is not in the document]
Comment (aphonpra-oai) [reply codex-rs/core-plugins/src/remote.rs]: Addressed in commit 6be86c2. 5xx responses now use the same bounded Retry-After gate as 429s, so long or unparsable Retry-After values skip the immediate retry instead of retrying inside the server-requested delay.

File: codex-rs/core-plugins/src/remote.rs [hunk 3]
```diff
@@ -1840,25 +1844,143 @@ async fn send_and_decode<T: for<'de> Deserialize<'de>>(
...
-    if !status.is_success() {
-        return Err(RemotePluginCatalogError::UnexpectedStatus {
+    let retry_after = parse_retry_after_delay(response.headers());
+    let body = response.text().await.map_err(|source| {
+        RemotePluginCatalogAttemptError::new(RemotePluginCatalogError::Request {
             url: url.to_string(),
-            status,
-            body,
+            source,
+        })
```
[the remark this answers is not in the document]
Comment (aphonpra-oai) [reply codex-rs/core-plugins/src/remote.rs]: Addressed in commit 6be86c2. `send_and_decode_attempt` now captures Retry-After before reading the body and carries it through body-read failures, so retryable statuses with interrupted bodies still honor the bounded Retry-After behavior.

**Review by xl-openai — APPROVED** [review]

**Comment by xl-openai** [comment]
@aphonpra-oai in your PR description can you add more details on things like retry count / timeout settings / ...
